documentation/adr/0017-separate-absolute-and-relative-indexing-APIs-in-pixel-array.md

```diff
@@ -0,0 +1,70 @@
+[<-previous](./0016-use-double-array-in-memory-for-pixel-data.md) |
+next->
+
+# 17 - Separate absolute and relative indexing APIs in pixel array
+
+Date: 2020-Oct-15
+
+## Status
+
+Accepted
+
+## Context
+
+The pixel array within an SQW object can be too large to fit into memory.
+To avoid running out of memory, the object holding the pixel array can be
+file-backed.
+This means that only a "page" of the pixel array is loaded into memory at any
+one time.
+
+Therefore two possible ways to index into the pixel array exist:
+
+1. **Absolute index**:
+_The position of the pixel in the full, file-backed, pixel array_.
+
+2. **Relative index**:
+_The position of the pixel in the currently loaded page of pixel data._
+
+## Decision
+
+There will be two separate APIs for accessing data,
+these APIs distinguish between the two types of indexing.
+
+1. **Perform absolute indexing  using `get_` methods:**
+
+    Obtaining a subset of pixels or pixel data by absolute index will be
+    possible using a `get_` method.
+    For example, the following will retrieve pixels 100-200 by absolute index:
+
+    ```matlab
+    pixels.get_pixels(100:200)
+    ```
+
+    Similarly, to retrieve a range of data from particular pixel array fields:
+
+    ```matlab
+    pixels.get_data(100:200, {'signal', 'variance'})
+    ```
+
+2. **Perform relative indexing using attribute:**
+
+    Obtaining pixel data using an attribute will return just the data for the
+    currently cached page.
```

Comment:
The [`PixelData` help text](https://github.com/pace-neutrons/Horace/blob/efb681809a12a590609eccec62dd1f94632f1612/horace_core/sqw/PixelData/%40PixelData/PixelData.m#L19) says
> No pixel data will be loaded from the file on construction.

Will the relative indexing silently load a page of data if the `PixelData` object's cache is empty? Or would it be better to issue an error?

Reply:
Data will be silently loaded when a getter is called

Reply:
Checking I have understood correctly. Harry has said that data will be silently loaded if a get_ method is called. Greg has asked if data will be silently loaded if the attribute accessed directly by index range e.g signal(1:100). Guessing that this will not load a page but produce an error if signal etc have not been loaded. These seem to be two different scenarios, maybe need separate texts.

Reply:
Sorry, I probably wasn't clear enough. By "getter" I meant a `get_` method or an attribute (the attributes are implemented in Matlab as `get.<attribute>` functions, so they are technically getters as well).

Constructing a `PixelData` object from a file, then calling `pix.signal(1:100)` will get you the signal at indices 1 to 100 from page 1. But the data is only loaded from file on the `pix.signal` call.